File: iOS/PanoAudioChat/PanoAudioChatSDK/Source/Sections/AudioRoom/PanoAudioRoomPresenter.m
//
//  PanoAudioRoomPresenter.m
//  PanoAudioChatSDK
//
//  Created by pano on 2021/7/19.
//

#import "PanoAudioRoomPresenter.h"
#import "PanoClientService.h"
#import "PanoMicQueueView.h"
#import "PanoMessageDefine.h"
#import "PanoClientService+OpenApi.h"
#import "PanoClientService+App.h"

@interface PanoAudioRoomPresenter ()
<PanoMicCollectionCellDelegate,
PanoUserDelegate,
PanoClientDelegate>

@end

@implementation PanoAudioRoomPresenter {
    BOOL isFirstReceiveMicInfo;
}

- (instancetype)initWithUserMode:(PanoUserMode)userMode {
    self = [super init];
    if (self) {
        isFirstReceiveMicInfo = true;
        dataSource = [[PanoAudioRoomDataSource alloc] init];
        dataSource.userMode = userMode;
        __weak typeof(self) weakSelf = self;
        dataSource.applyBlock = ^(NSArray<PanoMicInfo *> * _Nonnull micInfos, PanoCmdReason reason) {
            [weakSelf reloadApplyView];
            for (id item in micInfos) {
                [weakSelf rejectChat:item reason:reason];
            }
        };
        [PanoClientService.service addDelegate:self];
        [PanoClientService.service.userService addDelegate:self];
    }
    return self;
}

#pragma mark -- PanoMicCollectionCellDelegate

- (void)onMicButtonPressed:(PanoMicInfo *)micInfo {
    __weak typeof(self) weakSelf = self;
    if (dataSource.userMode == PanoAnchor) {
        if (micInfo.status == PanoMicNone) {
            [_delegate showInvitePage:micInfo];
        } else if (micInfo.isOnline) {
            BOOL muteFlag = true;
            PanoUserInfo *user = micInfo.user;
            if (user) {
                muteFlag = user.audioStatus == PanoUserAudio_Unmute;
            }
            [_delegate showkickMicActionSheet:muteFlag handler:^(PanoActionSheetType type) {
                if (type == PanoActionKillUser) {
                    [weakSelf killUser:micInfo];
                } else if (type == PanoActionMuteUser) {
                    [weakSelf muteUser:micInfo];
                } else if (type == PanoActionUnmuteUser) {
                    [weakSelf unmuteUser:micInfo];
                }
            }];
        }
    } else {
        if (dataSource->myMicInfo.status == PanoMicConnecting) {
            // 您正在连麦中，无法申请上麦
            return;
        }
        
        // 如果当前麦位上有人，或者正在申请该麦位
        if ((micInfo.isOnline || micInfo.status == PanoMicConnecting) &&
            micInfo.userId != dataSource->myMicInfo.userId ){
            return;
        }
        if (dataSource->myMicInfo.status == PanoMicNone) {
            __block PanoMicInfo *myMicInfo = dataSource->myMicInfo;
            [_delegate showApplyAlert:^(BOOL success) {
                myMicInfo.status = PanoMicConnecting;
                myMicInfo.order = micInfo.order;
                if (myMicInfo != nil) {
                    [self applyChat:myMicInfo];
                }
                // 已申请上麦，等待通过...
            }];
        } else if (micInfo.status == PanoMicConnecting) {
            NSLog(@"%@ 正在申请该麦位", micInfo.user.userName);
        } else if  (micInfo.isOnline) {
            //申请下麦
            [_delegate exitChat];
        }
    }
}



- (void)onUserAdded:(PanoUserInfo *)user {
    [_delegate updateNaviTitleView];
}

- (void)onUserRemoved:(PanoUserInfo *)user {
    [self remove:user];
    [_delegate updateNaviTitleView];
    [_delegate reloadMicQueueView];
}

- (void)onUserPropertyChanged:(PanoUserInfo *)user {
    [self reloadView];
    PanoUserInfo *host = PanoClientService.service.userService.host;
    if (host) {
        [_delegate updateHeaderView:host];
    }
}

- (void)onUserAudioLevelChanged {
    [self onUserPropertyChanged:nil];
}

#pragma mark --PanoClientDelegate

- (void)showExitRoomAlert:(NSString *)message {
    NSString *msg = message != nil ?  message : NSLocalizedString(@"加入房间失败", nil);
    [_delegate showExitAlert:msg handler:^(BOOL success) {
    }];
}

- (void)onPropertyChanged:(NSDictionary *)value forKey:(NSString *)key {
    if (![key isEqualToString:PanoAllMicKey]) {
        return;
    }
    PanoMicMessage *cmd = [[PanoMicMessage alloc] initWithDict:value];
    if (cmd == nil) {
        return;
    }
    
    if (PanoClientService.service.userService.isHost) {
        if (isFirstReceiveMicInfo) {
            isFirstReceiveMicInfo = false;
            [dataSource updateMicArrayWithMicInfos:cmd.data];
        }
        return;
    }
    BOOL found = false;
    for (PanoMicInfo *item in cmd.data) {
        if (item.userId == dataSource->myMicInfo.userId) {
            found = true;
        }
    }
    if (!found && dataSource->myMicInfo.isOnline ) {
        [PanoClientService.service.audioService muteAudio];
        [dataSource resetMyMic];
    }
    [dataSource updateMicArrayWithMicInfos:cmd.data];
    [self reloadView];
    
    switch (dataSource->myMicInfo.status) {
        case PanoMicFinished:
            [PanoClientService.service.audioService unmuteAudio];
            break;
        default:
            [PanoClientService.service.audioService muteAudio];
            break;
    }
}

- (void)reloadView {
    [_delegate reloadControlView];
    [_delegate reloadMicQueueView];
}

- (void)reloadApplyView {
    [_delegate updateApplyChatView:dataSource.allApplicants];
}

/// 更新所有的麦位状态
- (void)updateAllMicStatus {
    NSDictionary *value = [[[PanoMicMessage alloc] initWithCmd:allMic data:dataSource.allMics reason:ok] toDictionary];
    [PanoClientService.service setProperty:value forKey:PanoAllMicKey];
}

- (void)onReceivedMessage:(NSDictionary<NSString *,id> *)message fromUser:(UInt64)userId {
    __weak typeof(self) weakSelf = self;
    PanoCmdMessage *cmdMsg = [[PanoCmdMessage alloc] initWithDict:message];
    if (cmdMsg && cmdMsg.cmd == closeRoom) {
        [PanoClientService.service uploadlogsAutomatically];
        [_delegate showExitAlert:nil handler:^(BOOL success) { }];
        [self leaveRoom];
    }
    PanoMicMessage *micMsg = [[PanoMicMessage alloc] initWithDict:message];
    if (!micMsg) {
        return;
    }
    PanoMicInfo *micInfo = [micMsg.data firstObject];
    if (!micInfo) {
        return;
    }
    PanoMicInfo *curMic = [dataSource micInfo:micInfo.order];
    if (!curMic) {
        return;
    }
    void(^rejectBlock)(PanoCmdReason reason) = ^(PanoCmdReason reason){
        PanoMicInfo *tempMicInfo = [[PanoMicInfo alloc] initWithMicInfo:micInfo];
        tempMicInfo.userId = userId;
        [self rejectChat:tempMicInfo reason:reason];
    };
    PanoUserInfo *fromUser = [PanoClientService.service.userService findUserWithId:userId];
    switch (micMsg.cmd) {
            // 收到发送申请上麦的消息
        case applyChat:
        {
            if (![PanoClientService.service.userService isHost]) {
                return;
            }
            if (curMic.isOnline) {
                rejectBlock(occupied);
                return;
            }
            micInfo.status = PanoMicConnecting;
            [dataSource appendApplyUser:micInfo];
            [self reloadApplyView];
        }
            break;
            // 收到主播同意上麦的消息
        case acceptChat:
        {
            if (dataSource->myMicInfo.userId == micInfo.userId) {
                [_delegate cancelApplyAlert];
                [self reloadView];
                [PanoClientService.service.audioService unmuteAudio];
            }
        }
            break;
            // 收到主播拒绝上麦的消息
        case rejectChat:
        {
            if (dataSource->myMicInfo.userId == micInfo.userId) {
                [_delegate cancelApplyAlert];
                NSString *msg = nil;
                switch (micMsg.reason) {
                    case timeout:
                        msg = @"主播长时间未同意你的申请上麦";
                        break;
                    case occupied:
                        msg = @"当前麦位被占用";
                        break;
                    default:
                        msg = @"主播拒绝了你的申请上麦";
                        break;
                }
                [_delegate showRejectedToast:[NSString stringWithFormat:NSLocalizedString(@"%@", nil), msg]];
                [dataSource resetMyMic];
            }
        }
            break;
            // 收到观众拒绝主播邀请消息
        case rejectInvite: {
            NSString *msg = [NSString stringWithFormat:@"'%@'%@", fromUser.userName, micMsg.reason == timeout ? NSLocalizedString(@"长时间未接受你的邀请", nil) :
                             NSLocalizedString(@"拒绝了你的邀请", nil)];
            [_delegate showToast:msg];
        }
            break;
        case cancelChat: {
            if (fromUser != nil) {
                [self remove:fromUser];
            }
        }
            
            break;
        case stopChat: {
            [dataSource updateMicArrayWithMicInfo:[[PanoMicInfo alloc] initWithStatus:PanoMicNone userId:0 order:micInfo.order]];
            [self updateAllMicStatus];
            [self reloadView];
        }
            break;
        case killUser: {
            if (dataSource->myMicInfo.userId == micInfo.userId) {
                [PanoClientService.service.audioService muteAudio];
                [dataSource resetMyMic];
                NSString *msg = [NSString stringWithFormat:@"'%@'%@", fromUser.userName, NSLocalizedString(@"您已被主播请下麦位", nil)];
                [_delegate showToast:msg];
            }
        }
            break;
            // 收到观众接受主播邀请消息
        case acceptInvite: {
            if (curMic.isOnline) {
                rejectBlock(occupied);
                return;
            }
            micInfo.status = PanoMicFinished;
            [dataSource updateMicArrayWithMicInfo:micInfo];
            [dataSource removeApplyUser:micInfo deleteAll:true];
            [self updateAllMicStatus];
            [self reloadView];
        }
            break;
        case inviteUser: {
            [_delegate showInvitedAlert:NSLocalizedString(@"主播邀请您上麦", nil) handler:^(BOOL success, PanoCmdReason reason) {
                if (success) {
                    [weakSelf acceptInvite:micInfo];
                } else {
                    [weakSelf rejectInvite:micInfo reason:reason];
                }
            }];
        }
            break;
        default:
            break;
    }
}

- (void)remove:(PanoUserInfo *)user {
    if (PanoClientService.service.userService.isHost) {
        BOOL contain = [dataSource containUser:user];
        [dataSource removeApplyUser:user];
        [dataSource removeMicUser:user];
        [self reloadApplyView];
        if (contain) {
            [self updateAllMicStatus];
            [self reloadView];
        }
    }
}

#pragma mark -- Public Methods

- (void)stopMyChat {
    [self stopChat:dataSource->myMicInfo];
    [dataSource resetMyMic];
}

- (void)stopChat:(PanoMicInfo *)info {
    [PanoClientService.service.signalService sendStopChatMsg:info];
}

- (void)accpetChat:(PanoMicInfo *)info {
    [dataSource removeApplyUser:info deleteAll:true];
    PanoMicInfo *tempInfo = [[PanoMicInfo alloc] initWithMicInfo:info];
    tempInfo.status = PanoMicFinished;
    [dataSource updateMicArrayWithMicInfo:tempInfo];
    [self updateAllMicStatus];
    [self reloadView];
    [self reloadApplyView];
    [PanoClientService.service.signalService sendAcceptChatMsg:info];
}

- (void)rejectChat:(PanoMicInfo *)info reason:(PanoCmdReason)reason {
    [dataSource removeApplyUser:info deleteAll:false];
    [self reloadApplyView];
    [PanoClientService.service.signalService sendRejectChatMsg:info reason:reason];
}

- (void)closeRoomMsg {
    [PanoClientService.service.signalService sendCloseRoomMsg];
}

- (void)applyChat:(PanoMicInfo *)info {
    [PanoClientService.service.signalService sendApplyChatMsg:info];
    [self performSelector:@selector(checkMyApplyStatus) withObject:nil afterDelay:PanoMsgExpireInterval];
}

- (void)checkMyApplyStatus {
    if (dataSource->myMicInfo.status == PanoMicConnecting) {
        [_delegate cancelApplyAlert];
        [_delegate showRejectedToast:NSLocalizedString(@"主播长时间未同意你的申请上麦", nil)];
        [dataSource resetMyMic];
    }
}

- (void)cancelChat {
    [NSObject cancelPreviousPerformRequestsWithTarget:self selector:@selector(checkMyApplyStatus) object:nil];
    [PanoClientService.service.signalService sendCancelChatMsg:dataSource->myMicInfo];
    PanoMicInfo *tempInfo = [[PanoMicInfo alloc] init];
    tempInfo.userId = PanoClientService.service.userId;
    dataSource->myMicInfo = tempInfo;
}

- (void)muteUser:(PanoMicInfo *)info  {
    PanoMicInfo *temp = [[PanoMicInfo alloc] initWithMicInfo:info];
    temp.status = PanoMicFinishedMuted;
    [dataSource updateMicArrayWithMicInfo:temp];
    [self updateAllMicStatus];
}

- (void)unmuteUser:(PanoMicInfo *)info  {
    PanoMicInfo *tempInfo = [[PanoMicInfo alloc] initWithMicInfo:info];
    tempInfo.status = PanoMicFinished;
    [dataSource updateMicArrayWithMicInfo:tempInfo];
    [self updateAllMicStatus];
}

- (void)joinAudioRoom:(void(^)(bool))completion {
    //    clientService = PanoClientService.service()
    //    clientService.joinAudioRoom(config: clientService.config) { [weak self] (flag) in
    //        guard self = self else { return completion(false) }
    //        if (clientService.config.userMode == PanoUserMode.anchor {
    //            user = PanoUserInfo(userId: clientService.config.userId, userName: clientService.config.userName)
    //            self. _delegate updateHeaderView(host: user)
    //        }
    //        completion(flag)
    //        print("joinAudioRoom")
    //    }
}

- (void)leaveRoom {
    [_delegate cancelApplyAlert];
    _delegate = nil;
    [PanoClientService.service leaveAudioRoom];
}
/// 主播踢掉某个麦位的用户
- (void)killUser:(PanoMicInfo *)info {
    [dataSource removeMicArray:info];
    [self updateAllMicStatus];
    [PanoClientService.service.signalService sendkillMsg:info];
    [self reloadView];
}

/// 接受主播邀请
- (void)acceptInvite:(PanoMicInfo *)info {
    [PanoClientService.service.signalService sendAcceptInviteMsg:info];
}

/// 拒绝主播邀请
- (void)rejectInvite:(PanoMicInfo *)info reason:(PanoCmdReason)reason {
    [PanoClientService.service.signalService sendRejectInviteMsg:info reason:reason];
}

#pragma mark -- PanoAudioPlayerDelegate

- (void)didChooseNextAction {
    [PanoClientService.service.playerService chooseNextAction];
}

- (void)didStartPlayAction:(BOOL)pause {
    [PanoClientService.service.playerService startPlayAction:pause];
}

- (void)didShowMoreAction {
    [_delegate showAudioPanelView];
}

@end

File: iOS/PanoAudioChat/PanoAudioChat/Classes/Sections/AudioRoom/PanoAudioRoomPresenter.swift
//
//  PanoAudioRoomPresenter.swift
//  PanoAudioChat
//
//  Created by wenhua.zhang on 2020/10/29.
//

import UIKit

public enum PanoActionSheetType : Int {
    case muteUser = 0
    case unmuteUser = 1
    case killUser = 2
}

protocol PanoAudioRoomInterface : NSObjectProtocol {
    
    func showInvitePage(micInfo: PanoMicInfo);
    
    func showApplyAlert(handler: @escaping (_ success: Bool) -> Void);
    
    func cancelApplyAlert();
    
    func showExitAlert(message: String?, handler: @escaping (Bool) -> Void);
    
    func showkickMicActionSheet(mute: Bool, handler: @escaping (PanoActionSheetType) -> Void);
    
    func updateNaviTitleView();
    
    func updateHeaderView(host: PanoUser);
    
    func showInvitedToast();
    
    func showToast(message: String?);

    func showInvitedAlert(message: String?, handler: @escaping (Bool, PanoCmdReason) -> Void);
    
    func showRejectedToast(message: String?);
    
    func updateApplyChatView(applyUsers: [PanoMicInfo]);
    
    func reloadMicQueueView();
    
    func showAudioPanelView();
    
    func reloadControlView();
    
    func exitChat();
}

class PanoAudioRoomPresenter : NSObject {

    var dataSource : PanoAudioRoomDataSource!
    
    weak var delegate: PanoAudioRoomInterface?
    
    var isFirstReceiveMicInfo = true;
    
    init(userMode: PanoUserMode) {
        
        super.init()
        
        dataSource = PanoAudioRoomDataSource()
        dataSource.userMode = userMode
        dataSource.applyArrayDidTrimmed = { [weak self] (micInfos: [PanoMicInfo], reason: PanoCmdReason) in
            self?.reloadApplyView()
            for item in micInfos {
                self?.rejectChat(info: item, reason: reason)
            }
        }
        
        let clientService = PanoServiceManager.service(type: .client) as! PanoClientService
        clientService.delegates.addDelegate(self)
        
        let userService = PanoServiceManager.service(type: .user) as! PanoUserService
        userService.delegates.addDelegate(self)
        
    }
    
    func clientService() ->PanoClientService {
        let clientService =  PanoServiceManager.service(type: .client) as! PanoClientService
        return clientService;
    }
    
    deinit {
        print(self, "deinit")
    }
}


extension PanoAudioRoomPresenter: PanoMicCollectionCellDelegate {
    
    func onMicButtonPressed(micInfo: PanoMicInfo) {
        
        if dataSource.userMode == .anchor {
            if micInfo.status == PanoMicStatus.none {
                delegate?.showInvitePage(micInfo: micInfo)
            } else if micInfo.isOnline() {
                var muteFlag = true
                if let user = micInfo.user {
                    muteFlag = user.audioStatus == .unmute
                }
                delegate?.showkickMicActionSheet(mute: muteFlag, handler: { [weak self] (type) in
                    if type == .killUser {
                        self?.killUser(info: micInfo)
                    } else if type == .muteUser {
                        self?.muteUser(info: micInfo)
                    } else if type == .unmuteUser {
                        self?.unmuteUser(info: micInfo)
                    }
                })
            }
        } else {
            if dataSource.myMicInfo.status == PanoMicStatus.connecting {
                // 您正在连麦中，无法申请上麦
                return
            }
            
            // 如果当前麦位上有人，或者正在申请该麦位
            if (micInfo.isOnline() || micInfo.status == .connecting) &&
                micInfo.userId != dataSource.myMicInfo.userId {
                return
            }
            
            if dataSource.myMicInfo.status == PanoMicStatus.none {
                
                delegate?.showApplyAlert { [weak self] (flag) in
                    self?.dataSource.myMicInfo.status = .connecting
                    self?.dataSource.myMicInfo.order = micInfo.order
                    if let info = self?.dataSource.myMicInfo {
                        self?.applyChat(info: info)
                    }
                    // 已申请上麦，等待通过...
                    
                }
                
            } else if micInfo.status == PanoMicStatus.connecting {
                
                _  = "\(String(describing: micInfo.user?.userName)) 正在申请该麦位";
                
            } else if  micInfo.isOnline() {
                
                //申请下麦
                self.delegate?.exitChat()
            }
        }
    }
}


extension PanoAudioRoomPresenter : PanoUserDelegate {
    
    func onUserAdded(user: PanoUser) {
        delegate?.updateNaviTitleView()
    }
    
    func onUserRemoved(user: PanoUser) {
        remove(user: user)
        delegate?.updateNaviTitleView()
        delegate?.reloadMicQueueView()
    }
    
    func onUserPropertyChanged(user: PanoUser?) {
        reloadView()
        if let host = PanoUserService.service()?.host {
            delegate?.updateHeaderView(host: host)
        }
    }
    
    func onUserAudioLevelChanged() {
        onUserPropertyChanged(user: nil);
    }
}

extension PanoAudioRoomPresenter: PanoClientDelegate {
    
    func showExitRoomAlert(message: String?) {
        delegate?.showExitAlert(message: message ?? "加入房间失败") { (flag) in
        }
    }
    
    func onPropertyChanged(_ value: [String : Any], for key: String) {
        print("v:", value, "key:", key);
        guard key == PanoAllMicKey else {
            return
        }
        guard let cmd = PanoMicMessage(dict: value) else {
            return
        }
        guard !PanoUserService.isHost() else {
            if isFirstReceiveMicInfo {
                isFirstReceiveMicInfo = false
                dataSource.updateMicArray(micInfos: cmd.data as! [PanoMicInfo])
            }
            return
        }
        dataSource.updateMicArray(micInfos: cmd.data as! [PanoMicInfo])
        reloadView()
        let audioService = PanoServiceManager.service(type: .audio) as! PanoAudioService
        switch dataSource.myMicInfo.status {
        case .finished:
            audioService.unmuteAudio()
            break
        default:
            audioService.muteAudio()
        }
    }
    
    func reloadView() {
        delegate?.reloadControlView()
        delegate?.reloadMicQueueView()
    }
    
    func reloadApplyView() {
        delegate?.updateApplyChatView(applyUsers: dataSource.applyMicArray)
    }
    
    /// 更新所有的麦位状态
    func updateAllMicStatus() {
        let value = PanoMicMessage(cmd: .allMic, data: dataSource.totalMicArray).toDictionary()
        _ = clientService().setProperty(value: value, for: PanoAllMicKey)
    }
    
    func onMessageReceived(_ message: [String : Any], fromUser userId: UInt64) {
        print("_onMessageReceived", message, userId)
        if let cmdMsg = PanoCmdMessage(dict: message),
           cmdMsg.cmd == .closeRoom {
            PanoClientService.service().uploadlogsAutomatically()
            delegate?.showExitAlert(message: nil) { (_) in }
            self.leaveRoom()
        }
        guard let micMsg = PanoMicMessage(dict: message) else {
            return
        }
        guard var micInfo = (micMsg.data as? [PanoMicInfo])?.first else {
            return
        }
        guard let curMic = dataSource.micInfo(with: micInfo.order) else {
            return
        }
        let reject = { [weak self] (reason: PanoCmdReason) in
            var tempMicInfo = PanoMicInfo(info: micInfo);
            tempMicInfo.userId = userId
            self?.rejectChat(info: tempMicInfo, reason: reason)
        };
        
        let fromUser = PanoUserService.service()?.findUser(userId: userId)
        switch micMsg.cmd {
        case .applyChat: // 收到发送申请上麦的消息
            guard PanoUserService.isHost() else {
                return
            }
            guard !curMic.isOnline() else {
                reject(.occupied);
                return
            }
            micInfo.status = .connecting
            dataSource.appendApplyUser(with: micInfo)
            reloadApplyView()
            break
        case .acceptChat: // 收到主播同意上麦的消息
            if dataSource.myMicInfo.userId == micInfo.userId{
                delegate?.cancelApplyAlert()
                reloadView()
                PanoAudioService.service()?.unmuteAudio()
            }
            break
        case .rejectChat: // 收到主播拒绝上麦的消息
            if dataSource.myMicInfo.userId == micInfo.userId {
                delegate?.cancelApplyAlert()
                var msg: String? = nil;
                switch micMsg.reason {
                case .timeout:
                    msg = "主播长时间未同意你的申请上麦";
                    break
                case .occupied:
                    msg = "当前麦位被占用";
                    break
                case .ok:
                    msg = nil;
                }
                delegate?.showRejectedToast(message: msg?.localized())
                dataSource.resetMyMic()
            }
        case .rejectInvite: // 收到观众拒绝主播邀请消息
            var msg = "'\(String(describing: fromUser?.userName ?? ""))'" ;
            msg += micMsg.reason == .timeout ? "长时间未接受你的邀请".localized() :                "拒绝了你的邀请".localized();
            delegate?.showToast(message: msg);
            break
        case .cancelChat:
            if let fromUser = fromUser {
                remove(user: fromUser)
            }
        case .stopChat:
            dataSource.updateMicArray(micInfo: PanoMicInfo(status: .none, order: micInfo.order))
            updateAllMicStatus()
            reloadView()
            break
        case .killUser:
            if dataSource.myMicInfo.userId == micInfo.userId {
                PanoAudioService.service()?.muteAudio()
                dataSource.resetMyMic()
                delegate?.showToast(message: "'\(String(describing: fromUser?.userName ?? ""))'" + "您已被主播请下麦位".localized());
            }
            break
        case .acceptInvite: // 收到观众接受主播邀请消息
            guard !curMic.isOnline() else {
                reject(.occupied);
                return
            }
            micInfo.status = PanoMicStatus.finished
            dataSource.updateMicArray(micInfo: micInfo)
            dataSource.removeApplyUser(with: micInfo);
            updateAllMicStatus()
            reloadView()
            break
        case .inviteUser:
            delegate?.showInvitedAlert(message: "主播邀请您上麦".localized(), handler: { [weak self] (flag, reason) in
                if (flag) {
                    self?.acceptInvite(info: micInfo)
                } else {
                    self?.rejectInvite(info: micInfo, reason: reason)
                }
            })
        default:
            break
        }
    }
    
    func remove(user: PanoUser) {
        if PanoUserService.isHost() {
            let contain = dataSource.containUser(user:user)
            dataSource.removeApplyUser(with: user)
            dataSource.removeMicUser(user: user)
            reloadApplyView()
            if contain {
                updateAllMicStatus()
                reloadView()
            }
        }
    }
}


extension PanoAudioRoomPresenter {
    
    func stopChat() {
        stopChat(info: dataSource.myMicInfo)
        dataSource.resetMyMic()
    }
    
    func stopChat(info: PanoMicInfo) {
        _ = clientService().signalService?.sendStopChatMsg(micInfo: info)
    }
    
    func accpetChat(info: PanoMicInfo) {
        dataSource.removeApplyUser(with: info)
        var tempInfo = PanoMicInfo(info: info)
        tempInfo.status = PanoMicStatus.finished
        dataSource.updateMicArray(micInfo: tempInfo)
        updateAllMicStatus()
        reloadView()
        reloadApplyView()
        _ = clientService().signalService?.sendAcceptChatMsg(micInfo: info)
    }
    
    func rejectChat(info: PanoMicInfo, reason: PanoCmdReason = .ok) {
        dataSource.removeApplyUser(with: info, delete: false)
        reloadApplyView()
        _ = clientService().signalService?.sendRejectChatMsg(micInfo: info, reason: reason)
    }
    
    func closeRoomMsg() {
        _ = clientService().signalService?.sendCloseRoomMsg()
    }
    
    func applyChat(info: PanoMicInfo) {
        _ = clientService().signalService?.sendApplyChatMsg(micInfo: info)
        perform(#selector(checkMyApplyStatus), with: nil, afterDelay: TimeInterval(PanoMsgExpireInterval))
    }
    
    @objc func checkMyApplyStatus() {
        if dataSource.myMicInfo.status == .connecting {
            delegate?.cancelApplyAlert()
            delegate?.showRejectedToast(message: "主播长时间未同意你的申请上麦".localized())
            dataSource.resetMyMic()
        }
    }
    
    func cancelChat(info: PanoMicInfo) {
        var tempInfo = PanoMicInfo()
        tempInfo.userId = PanoClientService.service().userId
        dataSource.myMicInfo = tempInfo
        _ = clientService().signalService?.sendCancelChatMsg(micInfo: info)
    }
    
    func muteUser(info: PanoMicInfo)  {
        var temp = PanoMicInfo(info: info)
        temp.status = .finishedMuted
        dataSource.updateMicArray(micInfo: temp)
        updateAllMicStatus()
    }
    
    func unmuteUser(info: PanoMicInfo)  {
        var tempInfo = PanoMicInfo(info: info)
        tempInfo.status = .finished
        dataSource.updateMicArray(micInfo: tempInfo)
        updateAllMicStatus()
    }
    
    func joinAudioRoom(completion: @escaping (Bool) -> Void) {
        let clientService = PanoClientService.service()
        clientService.joinAudioRoom(config: clientService.config) { [weak self] (flag) in
            guard let self = self else { return completion(false) }
            if clientService.config.userMode == PanoUserMode.anchor {
                let user = PanoUser(userId: clientService.config.userId, userName: clientService.config.userName)
                self.delegate?.updateHeaderView(host: user)
            }
            completion(flag)
            print("joinAudioRoom")
        }
    }
    
    func leaveRoom() {
        delegate?.cancelApplyAlert()
        delegate = nil
        PanoClientService.service().leaveAudioRoom()
    }
    /// 主播踢掉某个麦位的用户
    func killUser(info: PanoMicInfo) {
        dataSource.removeMicArray(micInfo: info)
        updateAllMicStatus()
        _ = clientService().signalService?.sendkillMsg(micInfo: info)
        reloadView()
    }
    
    /// 接受主播邀请
    func acceptInvite(info: PanoMicInfo) {
        _ = clientService().signalService?.sendAcceptInviteMsg(micInfo: info)
    }
    
    /// 拒绝主播邀请
    func rejectInvite(info: PanoMicInfo, reason: PanoCmdReason = .ok) {
        _ = clientService().signalService?.sendRejectInviteMsg(micInfo: info, reason: reason)
    }
}


extension PanoAudioRoomPresenter : PanoAudioPlayerDelegate {
    
    func didChooseNextAction() {
        PanoPlayerService.service()?.chooseNextAction()
    }
    
    func didStartPlayAction(pause: Bool) {
        PanoPlayerService.service()?.startPlayAction(pause: pause)
    }
    
    func didShowMoreAction() {
        delegate?.showAudioPanelView()
    }
}



